### PR TITLE
Add watching UI

### PR DESCRIPTION
**Pull request by Arkitecht**
Adds watching indicator to discussion list view
Adds watch / unwatch buttons to discussion view
Adds ForumController tests

Screen Shot 2021-05-28 at 12 46 23 PM

Screen Shot 2021-05-28 at 12 47 12 PM

Screen Shot 2021-05-28 at 12 47 01 PM

### CODE REVIEW

File: src/Http/Controllers/ForumController.php [hunk 1]
```diff
@@ -12,12 +14,21 @@ class ForumController extends Controller
...
+        if (config('firefly.features.watchers') && $request->user()) {
+            $discussionQuery->leftJoin('discussion_user', function (JoinClause $join) use ($request) {
+                $join->on('discussion_user.discussion_id', '=', 'discussions.id')
+                    ->where('discussion_user.user_id', $request->user()->id);
+            })->select([
+                'discussions.*',
+                'discussion_user.user_id as is_being_watched',
+            ]);
+        }
+        $discussions = $discussionQuery->paginate(config('firefly.pagination.discussions'));
```
Comment (tomirons) [inline src/Http/Controllers/ForumController.php:31]: Here is something that's a bit cleaner. I'm also curious if we could put the same `watchers` relation on the discussion? That would clean this up even more IMO.
```suggestion
        $discussions = Discussion::query()
            ->when(config('firefly.features.watchers') && $user = $request->user(), function ($query) use ($user) {
                $query->leftJoin('discussion_user', function (JoinClause $join) use ($user) {
                    $join->on('discussion_user.discussion_id', '=', 'discussions.id')
                        ->where('discussion_user.user_id', $user->id);
                })->select([
                    'discussions.*',
                    'discussion_user.user_id as is_being_watched',
                ]);
            })
            ->orderBy('created_at', 'desc')
            ->paginate(config('firefly.pagination.discussions'));
```

File: tests/TestCase.php [hunk 2]
```diff
@@ -37,6 +37,15 @@ protected function setUp(): void
...
         $app['config']->set('firefly.api.enabled', true);
+        $app['config']->set('firefly.user', User::class);
+    }
+
+    /**
+     * Define environment for testing watchers.
+     */
+    protected function usesWatchers($app)
+    {
+        $app->config->set('firefly.features.watchers', true);
```
Comment (tomirons) [inline tests/TestCase.php:48]: Is there a better way to do this? Maybe do a call to `enableWatchersFeature` inside of the `setUp` methods of the tests that need it? I personally would rather not use `@define-env` in tests.

**Review by tomirons — CHANGES_REQUESTED** [review]

File: src/Http/Controllers/ForumController.php (hunk 1 above)
Comment (Arkitecht) [reply src/Http/Controllers/ForumController.php:31]: I updated the logic to incorporate your changes, and then further cleaned up by adding as a query scope on the Discussion model. 

Was not 100% sure what you meant by adding to the watchers relation on the discussion. 
We have a:
```php
public function watchers(): BelongsToMany
{
    return $this->belongsToMany(User::class);
}
```

File: resources/views/components/discussion-item.blade.php [hunk 3]
```diff
@@ -33,6 +33,10 @@
                 <span class="inline-flex items-center px-2.5 py-1 rounded-full border text-xs font-medium bg-white text-gray-700">
                     {{ $discussion->reply_count }} {{ $discussion->reply_count === 1 ? __('reply') : __('replies') }}
                 </span>
+
+                @if(config('firefly.features.watchers'))
```
Comment (AndreasElia) [inline resources/views/components/discussion-item.blade.php:37]: Space after `@if` to become `@if (config('firefly.features.watchers'))`

File: resources/views/components/discussion-watch-icon.blade.php [hunk 4]
```diff
@@ -0,0 +1,10 @@
+@props(['discussion'])
```
Comment (AndreasElia) [inline resources/views/components/discussion-watch-icon.blade.php:1]: New line below probs as spacer.

**Comment by AndreasElia** [comment]
I'm not entirely a fan of those icons as I don't believe they go for "watching" (more favouriting). Maybe we could go with the "eye" heroicons? https://heroicons.com/

**Comment by Arkitecht** [comment]
Made the requested changes, and updated the icon
Screen Shot 2021-06-01 at 4 19 57 PM

**Comment by AndreasElia** [comment]
Thanks @Arkitecht, much appreciated.

image

The vertical alignments of the eye looks a little off here. Also, do you have any thoughts or suggestions on making the icon a little more engaging as to what is active and what is inactively being watched? At the moment a user may get confused as to which is which 🤔

**Comment by Arkitecht** [comment]
Do we like this any better?
Fixes the alignment, and is a little more subtle, but may give the message better?

Screen Shot 2021-06-01 at 7 31 18 PM

Other possibilities:
Add eyelashes for the watching version (something like: https://www.popicon.com/products/watching-icon)
Or we can use one of the eye-off options from https://heroicons.com/ to indicate not-watching

Thoughts?

**Comment by AndreasElia** [comment]
Since they are simply icons to indicate the "watched" status, let's put in the standard "eye" icon for when it is being watched, and display no icon when it isn't being watched. This cleans up the UI a little bit too.

**Comment by Arkitecht** [comment]
That works for me, in that case I suggest moving it to the left side with the other possible icons, so the groups and replies stay in line

Original:
Screen Shot 2021-06-02 at 8 37 39 AM

Suggested:
Screen Shot 2021-06-02 at 8 42 31 AM

I also added in the tooltip.

Do you want me to combine the logic into the existing icon.blade.php file - instead of having the one just for the watching icon?

**Comment by AndreasElia** [comment]
@Arkitecht please combine the logic with the other icons, and move to the left to group with them too. Thank you.

File: src/Models/Discussion.php [hunk 5]
```diff
@@ -130,4 +132,17 @@ public function watchers(): BelongsToMany
...
+    {
+        $builder->when(config('firefly.features.watchers') && $user, function ($query) use ($user) {
+            $query->leftJoin('discussion_user', function (JoinClause $join) use ($user) {
+                $join->on('discussion_user.discussion_id', '=', 'discussions.id')
+                    ->where('discussion_user.user_id', $user->id);
+            })->select([
+                'discussions.*',
+                'discussion_user.user_id as is_being_watched',
+            ]);
+        });
```
Comment (tomirons) [inline src/Models/Discussion.php:146]: I found a much easier/cleaner way to do this! 😁 
```suggestion
        $builder->when(config('firefly.features.watchers'), function ($query) use ($user) {
            $query->withExists(['watchers as is_being_watched' => fn ($query) => $query->where('user_id', $user->id));
        });
```

**Review by tomirons — CHANGES_REQUESTED** [review]

File: src/Models/Discussion.php (hunk 5 above)
Comment (Arkitecht) [reply src/Models/Discussion.php:146]: That's really pretty!
Composer requires `"php": "^7.2 || ^8.0",` and short functions are available starting in 7.4, is that an issue?

Comment (AndreasElia) [reply src/Models/Discussion.php:146]: In that case, let's update composer to require 7.4 at a minimum.

Comment (Arkitecht) [reply src/Models/Discussion.php:146]: Ok, I can make that part of my next commit and this PR.
Thanks

Comment (tomirons) [reply src/Models/Discussion.php:146]: This package has 7.4 as a minimum.

https://github.com/AndreasElia/laravel-firefly/blob/f55ce93d0dd5e705172b8e3380c734da4596960c/composer.json#L22

Comment (Arkitecht) [reply src/Models/Discussion.php:146]: Weird, sorry about that, my IDE must have pulled up the wrong composer.json file.
All good, will send the last updates shortly!

**Comment by AndreasElia** [comment]
This is looking pretty darn awesome, thanks @Arkitecht. Anything else @tomirons?

**Review by tomirons — APPROVED** [review]

**Pull request merged**